Support OpenApi2.0 basePath attribute

tgiardina opened the pull request:
As described in Issue #159, this package currently ignores OpenApi2.0's basePath attribute. This PR aims to add support for it.

### Commits

Included are three commits, which:
1) Add tests that describe the problem.
2) Fix the problem via changes to [OpenApi2Spec.js ](https://github.com/RuntimeTools/OpenAPIValidators/blob/master/packages/chai-openapi-response-validator/lib/openapi-validator/lib/classes/OpenApi2Spec.js)and [satisfyApiSpec.js](https://github.com/RuntimeTools/OpenAPIValidators/blob/master/packages/chai-openapi-response-validator/lib/assertions/satisfyApiSpec.js). (I edited `satsifyApiSpec.js` so that error messages include `basePath` (if it exists) in the url. The error messages were confusing otherwise.)
3) Added myself to AUTHORS.md per the [contributing guidelines](https://github.com/RuntimeTools/OpenAPIValidators/blob/master/CONTRIBUTING.md#pull-requests).

### Needs Review

There are two main things that I think need review:

1) The problem can also be solved by considering `basePath` in OpenApi2.0 as analogous to `servers` in OpenApi3.0. We can then copy a lot of the functionality from `OpenApi3Spec.js` into `OpenApi2Spec.js` (e.g. `getServerUrls`, `getMatchingServerUrls`) and no longer need to edit `satisfyApiSpec.js`. I did not solve the problem this way because I think it would require a lot of thought re what belongs in `OpenApi2Spec.js` and `OpenApi3Spec.js` versus `AbstractOpenApiSpec.js`, and I did not feel like it was my place to make those type of decisions.
2) I only made changes to the [chai package](https://github.com/RuntimeTools/OpenAPIValidators/tree/master/packages/chai-openapi-response-validator) and so don't know if similar changes need to be made to the[ jest package](https://github.com/RuntimeTools/OpenAPIValidators/tree/master/packages/jest-openapi).

rwalle61 commented on packages/chai-openapi-response-validator/lib/openapi-validator/lib/classes/OpenApi2Spec.js:
```diff
@@ -2,10 +2,15 @@ const utils = require('../utils');
+class OpenApi2Spec extends AbstractOpenApiSpec { 
   findOpenApiPathMatchingPathname(pathname) {
+    const { basePath } = this.spec;
+    if(basePath && !pathname.startsWith(basePath)) {
```
if you run our mutation tests, you'll see this isn't needed to pass our tests ;) see the Travis logs https://github.com/RuntimeTools/OpenAPIValidators/pull/160/files#diff-585a46b55b8d80d190a56a2a5ff41a8bR47

So either our tests are insufficient, or this code is unnecesary (you choose, you're the end user!)

tgiardina replied:
Good catch. This is necessary, otherwise a spec with `basePath: /test` and path `/responseBody/string` will accept a path of `/responseBody/string`. My new commit "Updated basePath test" now tests exactly this situation. 

Also, I'm not familiar with mutation testing, so I may be missing stuff on that front. Will try to do a bit of research tonight!

rwalle61 replied:
Thanks for adding the test! Mutation testing is just a way of checking whether you have more code than necessary to pass your tests. We use it here to ensure both that we write no more code than we need to, and that our tests fully cover all the features we want to provide

rwalle61 commented on packages/chai-openapi-response-validator/lib/openapi-validator/lib/classes/OpenApi2Spec.js:
```diff
@@ -2,10 +2,15 @@ const utils = require('../utils');
+    if(basePath && !pathname.startsWith(basePath)) {
+      throw new ValidationError('PATH_NOT_FOUND');
+    }
+    const endpoint = pathname.replace(basePath, '');
```
I would probably reuse the code and naming conventions we use for OpenAPI 3, i.e.

```js
const pathnameWithoutBasePath = getPathnameWithoutBasePath(basePath, pathname),
    const openApiPath = utils.findOpenApiPathMatchingPossiblePathnames(
      [pathnameWithoutBasePath],
...
```
(`endpoint` can mean many things in a package about OpenAPI paths)

tgiardina replied:
Agreed. Changed `endpoint` to `pathnameWithoutBasePath` in the commit `Changed variable name for consistency`

rwalle61 commented on packages/chai-openapi-response-validator/test/assertions/satisfyApiSpec/openapi2SpecsDefineBasePath.test.js:
```diff
@@ -0,0 +1,97 @@
+      expect(assertion).to.throw(
+        "expected res to satisfy a '200' response defined for endpoint 'GET nonExistentServer/test/responseBody/string' in your API spec" +
+        "\nres had request path 'nonExistentServer/test/responseBody/string', but your API spec has no matching path" +
+                '\n\nPaths found in API spec: /test/responseBody/string'
```
in our OpenAPI 3 serversDefined tests, you'll see that when we can't find a matching OpenAPI path, but we can find a `server` / `basePath`, we say `Servers found in API spec: ... `

As a user of OpenAPI 2, would you find it useful to see something like `basePath found (/test) but matches no basePath + serverURL combinations`?

That may be a bigger task, so I don't mind if we/I do that in a later PR

tgiardina replied:
Good question! I don't think the longer message adds much utility in this case. If you have multiple `servers`, you need to explicitly think about which one you're using. But with `basePath`, you're only adding it so you don't have to type it over and over again -- because you don't want to explicitly think about it. So I think it's fine to just say the URL as a whole is no good.

rwalle61 replied:
Thanks for your reply, I agree.

Relatedly, do you think there's any risk that someone reading the error message
```
Paths found in API spec: /test/responseBody/string'
```
will be confused because they only have the `responseBody/string` path in the paths section of their spec?

Perhaps it's up to them to realise that we are combining their `basePath` with each path. Do you think we should provide a hint? (I think it's probably beyond the responsibility of this package).

rwalle61 replied:
In fact, this decision affects OpenAPI 3, because the OpenAPI 2 `basePath` is similar to the OpenAPI 3 `server` URLs. I think the error message
```
Paths found in API spec: /test/responseBody/string'
```
should be consistent between OpenAPI 2 and 3, so let's decide now for both OpenAPI 2 and 3.

An OpenAPI 2 spec can have only one `basePath`. Suppose in the `Paths found in API spec` error message, we list paths without distinguishing `basePath` (`/test`) and the 'endpoint' paths (`responseBody/string`). A user who defines 10 endpoint paths in their OpenAPI 2 spec will see 10. User is happy.

But an OpenAPI 3 spec can have many server URLs. So suppose in `Paths found in API spec` error message, we list paths without distinguishing server URLs and 'endpoint' paths. A user who defines 10 endpoint paths in their OpenAPI 3 spec will see 10 `Paths found in API spec` if they have 1 server URL. But if they have 5 server URLs, they will see 50 `Paths found in API spec`.

So, questions for users @tgiardina and @JonnySpruce (if you're interested):
- for OpenAPI 3, should we list all 50 paths, or instead list 10 paths and then list 5 servers?
- must OpenAPI 2 be consistent with OpenAPI 3 on this?
- (what is least confusing for the user?)

tgiardina replied:
I'm not sure we can keep the error messages consistent, because I feel strongly about two points:

1) In your OpenApi 3 example, we should list 10 paths instead of 50 paths.
2) Users using OpenApi 2 should not have to decipher what `servers` are when reading an error message.

My guess is that everyone will agree on Point 1. I think the 50 paths option will obfuscate the error message quite a bit. Some internal code consistency is not worth the price.

Re Point 2, I think there is a lot of software that doesn't support OpenApi 3 yet (citation needed?) and so users who have only ever used OpenApi 2 should be treated as first class citizens. Part of being a first class citizen is getting errors messages specifically tailored for you to understand. And for example, someone who has never used OpenApi 3 may not be able to quickly and easily decipher:

```
'/basePath/nonExistentEndpointPath' matches servers [ '/basePath' ] but no <server/endpointPath> combinations
       
      -Servers found in API spec: /basePath`
```

If we can somehow revise the above error message so that it's consistent and adheres to both points above, let's do that. Otherwise, I think we should sacrifice a little bit of consistency for a whole lot of clarify.

rwalle61 replied:
Okay, I agree with both points. Would these error messages satisfy them for OpenAPI 2 and 3, while keeping the two consistent?

OpenAPI 2:

When request path does not match your `basePath`:
```
expected res to satisfy a '200' response defined for endpoint 'GET /wrongBasePath' in your API spec

res had request path '/wrongBasePath', but your API spec has basePath '/basePath'
```

When request path matches your `basePath` but no endpoint paths:
```
expected res to satisfy a '200' response defined for endpoint 'GET basePath/nonExistentEndpointPath' in your API spec

res had request path '/basePath/nonExistentEndpointPath' but your API spec has no matching path

Paths found in API spec: /endpointPath

'/basePath/nonExistentEndpointPath' matches basePath `/basePath` but no <basePath/endpointPath> combinations
```


OpenAPI 3:

When request path does not match any of your `server`s (this removes the `Paths found in API spec` lines from the existing message):
```
expected res to satisfy a '200' response defined for endpoint 'GET /nonExistentServer' in your API spec

res had request path '/nonExistentServer', but your API spec has no matching servers

Servers found in API spec: /server1, /server2, /server3
```

When request path matches some of your `server`s but no endpoint paths:
```
expected res to satisfy a '200' response defined for endpoint 'GET /server1/nonExistentEndpointPath' in your API spec

res had request path '/server1/nonExistentEndpointPath' but your API spec has no matching path

Paths found in API spec: /test

'/server1/nonExistentEndpointPath' matches servers '/server1' but no <server/endpointPath> combinations

Servers found in API spec: /server1, /server2, /server3
```

JonnySpruce replied:
@rwalle61, I think the proposal above looks good to me, the only thing I would add potentially change is the OpenAPI 3 no endpoint paths message. I would simply remove the additional information about servers found, since you've already matched one and so that is probably information you don't need, and also keeps the structure consistent with OpenAPI 2:

```
expected res to satisfy a '200' response defined for endpoint 'GET /server1/nonExistentEndpointPath' in your API spec

res had request path '/server1/nonExistentEndpointPath' but your API spec has no matching path

Paths found in API spec: /test

'/server1/nonExistentEndpointPath' matches servers '/server1' but no <server/endpointPath> combinations
```

tgiardina replied:
@rwalle61, The proposed solution messages look good to me! I also agree with everything @JonnySpruce said! Although I will add that I don't think the additional `servers` information will confuse anyone. It's just sorta weird and unnecessary.

Let me know when you've solidified your choice, and I will wrap up this PR!

rwalle61 replied:
Thanks @JonnySpruce and @tgiardina, let's go with those messages (removing the `servers` info for OpneAPI 3)!

rwalle61 reviewed (commented):
Thanks very much for this! Here's a quick initial review, and I'll give a proper look when I can.

The general approach I agree with (as laid out in my comment https://github.com/RuntimeTools/OpenAPIValidators/issues/159#issuecomment-690618514), so I think any requested changes are likely to be minor :)

As for jest-openapi, let's get this finished and merged for chai before adding to jest

rwalle61 commented:
Thanks for the updates! I'm away this weekend so may not be able to review properly until next week, hope that's okay. Hopefully we can merge next week

tgiardina commented:
No rush!

rwalle61 commented on commonTestResources/exampleOpenApiFiles/valid/openapi2DefineBasePath.yml:
```diff
@@ -0,0 +1,17 @@
+paths:
+  /responseBody/string:
+    get:
+      produces:
```
nit: if we can remove
```js
produces:
  - application/json
    parameters: []
```
let's do so, to keep the test as minimal as possible

tgiardina replied:
Fix in [this commit](https://github.com/RuntimeTools/OpenAPIValidators/pull/160/commits/efa5fbe049f8b015818cb4412fb8a9235031e15a).

rwalle61 commented on commonTestResources/exampleOpenApiFiles/valid/openapi2DefineBasePath.yml:
```diff
@@ -0,0 +1,17 @@
+        - application/json
+      parameters: []
+      responses:
+        '200':
```
```suggestion
        200:
```

tgiardina replied:
Fixed in [this commit](https://github.com/RuntimeTools/OpenAPIValidators/pull/160/commits/efa5fbe049f8b015818cb4412fb8a9235031e15a).

rwalle61 commented on packages/chai-openapi-response-validator/test/assertions/satisfyApiSpec/openapi2SpecsDefineBasePath.test.js:
```diff
@@ -0,0 +1,97 @@
+  before(() => {
+    chai.use(chaiResponseValidator(path.resolve(
+      '../../commonTestResources/exampleOpenApiFiles/valid/'
+         + 'openapi2DefineBasePath.yml'
```
```suggestion
         , 'openapi2DefineBasePath.yml'
```
(and then put the comma on the preceding line)

tgiardina replied:
This was resolved by the ["Refactored basePath test " commit](https://github.com/RuntimeTools/OpenAPIValidators/pull/160/commits/126d6dd0ba46e327f8ec1bd35a2b837b319d98f6).

rwalle61 commented on packages/chai-openapi-response-validator/test/assertions/satisfyApiSpec/openapi2SpecsDefineBasePath.test.js:
```diff
@@ -0,0 +1,97 @@
+      expect(assertion).to.throw(AssertionError, '');
+    });
+  });
+  describe('res.req.path does not include the basePath', () => {
```
```suggestion
  describe('res.req.path does not match the basePath', () => {
```

tgiardina replied:
This was resolved by the ["Refactored basePath test " commit](https://github.com/RuntimeTools/OpenAPIValidators/pull/160/commits/126d6dd0ba46e327f8ec1bd35a2b837b319d98f6).

rwalle61 commented on packages/chai-openapi-response-validator/lib/assertions/satisfyApiSpec.js:
```diff
@@ -44,7 +44,7 @@ function getExpectedResToSatisfyApiSpecMsg(
       `\n\nexpected res to satisfy a '${status}' response defined for endpoint '${endpoint}' in your API spec` +
       `\nres had request path '${requestPath}', but your API spec has no matching path` +
-      `\n\nPaths found in API spec: ${openApiSpec.paths().join(', ')}`;
+      `\n\nPaths found in API spec: ${openApiSpec.paths().map(path => (openApiSpec.spec.basePath || "")+path).join(', ')}`;
```
could this OpenAPI 2- specific functionality be moved to `OpenAPI2Spec.js` by override its `paths()` method?

But first let's confirm what behaviour we want (see comment below)

rwalle61 reviewed (commented):
Looking good! We have 1 question on functionality to decide (see comment), but code-wise this is great :)

rwalle61 commented:
So [we've decided what error messages we want](https://github.com/RuntimeTools/OpenAPIValidators/pull/160#discussion_r490592622) for OpenAPI 2. 

I think this means we need to cover these test cases for OpenAPI 2:

```js
describe('spec has no basePath property'

    describe("res.req.path matches the default basePath ('/') and an endpoint path"
    // succeeds

    describe("res.req.path matches the default basePath ('/') but no endpoint paths"
    // expected res to satisfy a '200' response defined for endpoint 'GET basePath/nonExistentEndpointPath' in your API spec
    // res had request path '/nonExistentEndpointPath' but your API spec has no matching path
    // Paths found in API spec: /endpointPath

    // (omit: '/nonExistentEndpointPath' matches basePath `/` but no <basePath/endpointPath> combinations)

describe("spec has a basePath property"

    describe("res.req.path matches the basePath and an endpoint path"
    // succeeds

    describe("res.req.path does not match the basePath"
    // expected res to satisfy a '200' response defined for endpoint 'GET /wrongBasePath' in your API spec
    // res had request path '/wrongBasePath', but your API spec has basePath '/basePath'

    describe("res.req.path matches the basePath but no endpoint paths"
    // expected res to satisfy a '200' response defined for endpoint 'GET basePath/nonExistentEndpointPath' in your API spec
    // res had request path '/basePath/nonExistentEndpointPath' but your API spec has no matching path
    // Paths found in API spec: /endpointPath
    // '/basePath/nonExistentEndpointPath' matches basePath `/basePath` but no <basePath/endpointPath> combinations
```

@tgiardina how does that sound? You already have half of those test cases, and most of the implementation. I think to `(omit: '/nonExistentEndpointPath' matches basePath `/` but no <basePath/endpointPath> combinations)` you could use [our pattern for handling this case for OpenAPI 3 servers](https://github.com/RuntimeTools/OpenAPIValidators/blob/48eddb49bd/packages/chai-openapi-response-validator/lib/assertions/satisfyApiSpec.js#L48)

tgiardina commented:
@rwalle61, I haven't forgotten about this! I will do my best to review your comments and wrap this PR up tomorrow! 

I didn't initially include the 'spec has no basePath property' tests because I figured it was tested elsewhere, but I'm happy to do so so that our tests are consistent between OpenApi2 and OpenApi3.

I'll let you know if I run into any problems, but I think it should be pretty smooth sailing now that we've figured out what error messages we want!

tgiardina commented:
@rwalle61 

- I added the additional test cases you mentioned in the ["Refactored basePath test " commit](https://github.com/RuntimeTools/OpenAPIValidators/pull/160/commits/126d6dd0ba46e327f8ec1bd35a2b837b319d98f6).
- I implemented the new OpenApi2 error messages in the ["Impl new error messages" commit](https://github.com/RuntimeTools/OpenAPIValidators/pull/160/commits).

However, this PR definitely still needs reviewing. In particular, there are two things I really don't feel good about:

1. I turned `satisfyApiSpec.js` into spaghetti implementing the new OpenApi2 error messages. In particular, I broke each error message down into a nested if-else, which is not very readable at all. I considered some other alternatives, but they all spiraled into larger refactors of `satisfyApiSpec.js` as a whole, and I don't think I'm qualified to make that type of decision. Instead, I [committed](https://github.com/RuntimeTools/OpenAPIValidators/pull/160/commits/a06aff3597b19cd67abd2118bfe3e8728a237b5d) my changes and will await feedback.
2. I also needed to add a `version` property (either `'2'` or `'3'`) to `AbstractOpenApiSpec`, which felt like it was beyond the scope of what these changes were supposed to be.

Thanks for patiently reviewing this PR with me! Please don't hesitate to ask any questions!

rwalle61 commented:
Thanks! Will have a look asap

rwalle61 commented on commonTestResources/exampleOpenApiFiles/valid/basePathVariations/basePathDefined.yml:
```diff
@@ -0,0 +1,14 @@
+        200:
+          description: Response body should be a string
+          schema:
+            type: string
```
nit: did you run prettier (`yarn format` / `npm run format`)? If yes, I will remember to update our prettier config to add new lines at end of files

tgiardina replied:
I did not run prettier, so I will make sure to do that for my next commits.

On a somewhat related note, you should probably update your [CONTRIBUTING](https://github.com/RuntimeTools/OpenAPIValidators/blob/master/CONTRIBUTING.md) documentation. A lot of the commands there are outdated (e.g., `npm run test:coverage` is no longer defined in `package.json`). You may also want to include a few words (or links) about mutation testing, because I don't think it's a very commonly used tool. 

I'm a big fan of this package, and I think you do a really great job managing it! Just want to make sure you keep your contributing guidelines up to date so that you don't have to spend any more time managing this repo than you already do!

rwalle61 replied:
- automate prettier https://github.com/RuntimeTools/OpenAPIValidators/pull/169
- update contributing.md https://github.com/RuntimeTools/OpenAPIValidators/pull/175

rwalle61 reviewed (commented):
Thanks, this PR is looking great! A few suggestions, and yes feel free to reopen a separate PR, because this one has swollen with discussion! (My apologies, but OpenAPI often turns out to be more complicated than I expect!)

tgiardina commented:
Closing this in favor of #172. There was a lot of good discussion here, but I think this PR has become a bit bloated as a result. I think #172 will be easier to decipher.
